jdk/test/sun/nio/cs/TestCharsetMapping.java

```diff
@@ -0,0 +1,640 @@
+                return false;
+            }
+            Matcher m = ptn.matcher("");
+            mappings = Files.lines(path)
+                .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                .map(ln -> parse(m))
+                .filter(e -> e.cp != UNMAPPABLE)  // non-mapping
+                .collect(Collectors.toList());
+            // xxx.nr
+            path = dir.resolve(clzName + ".nr");
+            if (Files.exists(path)) {
+                nr = Files.lines(path)
+                    .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                    .map(ln -> parse(m))
+                    .collect(Collectors.toMap(e -> e.bb, Function.identity()));
+            }
+            // xxx.c2b
+            path = dir.resolve(clzName + ".c2b");
+            if (Files.exists(path)) {
+                c2b = Files.lines(path)
+                    .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                    .map(ln -> parse(m))
+                    .collect(Collectors.toMap(e -> e.cp, Function.identity()));
+            }
+            return true;
+        }
+    }
+
+    private static Set<CharsetInfo> charsets(Path cslist, boolean sbcs) throws IOException {
+        Set<CharsetInfo> charsets = new LinkedHashSet<>();
+        Iterator<String> itr = Files.readAllLines(cslist).iterator();
+
+        while (itr.hasNext()) {
+            String line = itr.next();
+            if (line.startsWith("#") || line.length() == 0) {
+                continue;
+            }
+            String[] tokens = line.split("\\s+");
+            if (tokens.length < 5) {
+                continue;
+            }
+            CharsetInfo cs = new CharsetInfo(tokens[1], tokens[0]);
+            cs.hisName = tokens[2];
+            cs.pkgName = tokens[4];
+            if (sbcs) {
+                cs.type = "sbcs";
+            } else {
+                cs.type = tokens[3];
+            }
+            charsets.add(cs);
+        }
+        return charsets;
+    }
+
+    public static void main(String args[]) throws Exception {
+        Path dir = Paths.get(System.getProperty("test.src", ".") +
+                             "/../../../../make/data/charsetmapping");
+        if (!Files.exists(dir)) {
+            // not inside jdk repo, no mappings, exit silently
+            log.println("Nothing done, not in a jdk repo: ");
+            return;
+        }
+        if (args.length > 0 && "-v".equals(args[0])) {
+            // For debugging: java CoderTest [-v]
+            verbose = true;
+        }
+
+        int errors = 0;
+        int tested = 0;
+        int skipped = 0;
+        int known = 0;
+
+        // sbcs files have fewer fields and a set type of sbcs
+        Set<CharsetInfo> charsets = charsets(dir.resolve("sbcs"), true);
+        charsets.addAll(charsets(dir.resolve("extsbcs"), true));
+        charsets.addAll(charsets(dir.resolve("dbcs"), false));
```

Comment:
small nit, can you reorder and parse dbcs first, then comment, then the other two? So its clear that dbcs is excluded from comment

Reply:
Sure. I originally expected they'd need a separate method altogether and I guess I didn't clean it up completely when I realised I could simplify it to just a boolean.